### PR TITLE
add a catch-all route if needed

### PR DESCRIPTION
**Pull request by zhaohuabing**
fix: https://github.com/envoyproxy/gateway/issues/2507

### CODE REVIEW

File: internal/gatewayapi/translator.go [hunk 1]
```diff
@@ -201,12 +202,60 @@ func (t *Translator) Translate(resources *Resources) *TranslateResult {
...
 	return newTranslateResult(gateways, httpRoutes, grpcRoutes, tlsRoutes,
 		tcpRoutes, udpRoutes, clientTrafficPolicies, backendTrafficPolicies,
 		securityPolicies, xdsIR, infraIR)
 
 }
 
+// For filters without native per-route support, we need to add a catch-all route
+// to ensure that these filters are disabled for non-matching requests.
+// https://github.com/envoyproxy/gateway/issues/2507
+func addCatchAllRoute(xdsIR map[string]*ir.Xds) {
```
Comment (arkodg) [inline internal/gatewayapi/translator.go:217]: cc @mt-inside

File: internal/gatewayapi/translator.go [hunk 2]
```diff
@@ -201,12 +202,60 @@ func (t *Translator) Translate(resources *Resources) *TranslateResult {
...
+						len(r.QueryParamMatches) == 0 {
+						needCatchAllRoutePerHost[host] = false
+					}
+				}
+			}
+
+			for host, needCatchAllRoute := range needCatchAllRoutePerHost {
+				if needCatchAllRoute {
+					http.Routes = append(http.Routes, &ir.HTTPRoute{
+						Name: "catch-all",
```
[the remark this answers is not in the document]
Comment (zhaohuabing) [reply internal/gatewayapi/translator.go:244]: Does the route name need to be unique across all hostnames?

File: internal/gatewayapi/translator.go [hunk 3]
```diff
@@ -201,12 +202,60 @@ func (t *Translator) Translate(resources *Resources) *TranslateResult {
...
+						PathMatch: &ir.StringMatch{
+							Prefix: ptr.To("/"),
+						},
```
Comment (arkodg) [inline internal/gatewayapi/translator.go:247]: what happens when a prefix match of `/` already exists ? :)

Comment (arkodg) [reply internal/gatewayapi/translator.go:247]: are we relying on first match / last match for this ?

Comment (zhaohuabing) [reply internal/gatewayapi/translator.go:247]: > what happens when a prefix match of `/` already exists ? :)

We skip at line 242 

> Are we relying on first match / last match for this ?
Yes. The matches have been sorted, right?

File: internal/gatewayapi/translator.go [hunk 4]
```diff
@@ -203,12 +207,61 @@ func (t *Translator) Translate(resources *Resources) *TranslateResult {
...
+			// skip if there is already a catch-all route
+			for host := range needCatchAllRoutePerHost {
+				for _, r := range http.Routes {
+					if (r.Hostname == host &&
+						r.PathMatch != nil &&
+						r.PathMatch.Prefix != nil &&
+						*r.PathMatch.Prefix == "/") &&
+						len(r.HeaderMatches) == 0 &&
+						len(r.QueryParamMatches) == 0 {
+						needCatchAllRoutePerHost[host] = false
```
Comment (arkodg) [inline internal/gatewayapi/translator.go:241]: `delete` key maybe better here

File: internal/gatewayapi/translator.go [hunk 5]
```diff
@@ -203,12 +207,61 @@ func (t *Translator) Translate(resources *Resources) *TranslateResult {
...
+					}
+				}
+			}
+
+			for host, needCatchAllRoute := range needCatchAllRoutePerHost {
+				if needCatchAllRoute {
+					underscoredHost := strings.ReplaceAll(host, ".", "_")
+					http.Routes = append(http.Routes, &ir.HTTPRoute{
+						Name: fmt.Sprintf("%s/catch_all", underscoredHost),
```
Comment (arkodg) [inline internal/gatewayapi/translator.go:250]:
```suggestion
						Name: fmt.Sprintf("%s/catch-all-return-404", underscoredHost),
```

File: internal/gatewayapi/translator.go (hunk 4 above)
Comment (zhaohuabing) [reply internal/gatewayapi/translator.go:241]: Hah, it's good to know that it's safe to delete a key within range in golang. :-)

**Review by arkodg — APPROVED** [review]
LGTM thanks @zhaohuabing !

**Pull request merged**